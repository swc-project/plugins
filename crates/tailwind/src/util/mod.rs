pub use self::split_at_top_level_only::*;

mod split_at_top_level_only;
