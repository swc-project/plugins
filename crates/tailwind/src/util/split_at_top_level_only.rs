use crate::base::Candidate;

pub(crate) fn split_at_top_level_only(input: &str, separator: &str) -> Vec<Candidate> {}
