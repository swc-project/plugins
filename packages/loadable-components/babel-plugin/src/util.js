/* eslint-disable import/prefer-default-export */

export function getImportArg(callPath) {
  return callPath.get('arguments.0')
}
