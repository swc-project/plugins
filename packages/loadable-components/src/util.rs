use swc_core::ecma::ast::*;

pub(crate) fn get_import_arg(call: &CallExpr) -> &Expr {
    &call.args[0].expr
}
