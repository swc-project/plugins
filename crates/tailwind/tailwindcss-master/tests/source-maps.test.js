import postcss from 'postcss'
import { runWithSourceMaps as run, html, css, map } from './util/run'
import { parseSourceMaps } from './util/source-maps'

it('apply generates source maps', async () => {
  let config = {
    content: [
      {
        raw: html`
          <div class="with-declaration"></div>
          <div class="with-comment"></div>
          <div class="just-apply"></div>
        `,
      },
    ],
    corePlugins: { preflight: false },
  }

  let input = css`
    .with-declaration {
      background-color: red;
      @apply h-4 w-4 bg-green-500;
    }

    .with-comment {
      /* sourcemap will work here too */
      @apply h-4 w-4 bg-red-500;
    }

    .just-apply {
      @apply h-4 w-4 bg-black;
    }
  `

  let result = await run(input, config)
  let { sources, annotations } = parseSourceMaps(result)

  // All CSS generated by Tailwind CSS should be annotated with source maps
  // And always be able to point to the original source file
  expect(sources).not.toContain('<no source>')
  expect(sources.length).toBe(1)

  expect(annotations).toMatchSnapshot()
})

it('preflight + base have source maps', async () => {
  let config = {
    content: [],
  }

  let input = css`
    @tailwind base;
  `

  let result = await run(input, config)
  let { sources, annotations } = parseSourceMaps(result)

  // All CSS generated by Tailwind CSS should be annotated with source maps
  // And always be able to point to the original source file
  expect(sources).not.toContain('<no source>')
  expect(sources.length).toBe(1)

  expect(annotations).toMatchSnapshot()
})

it('utilities have source maps', async () => {
  let config = {
    content: [{ raw: `text-red-500` }],
  }

  let input = css`
    @tailwind utilities;
  `

  let result = await run(input, config)
  let { sources, annotations } = parseSourceMaps(result)

  // All CSS generated by Tailwind CSS should be annotated with source maps
  // And always be able to point to the original source file
  expect(sources).not.toContain('<no source>')
  expect(sources.length).toBe(1)

  expect(annotations).toStrictEqual(['2:4 -> 1:0', '2:4-23 -> 2:4-24', '2:4 -> 3:4', '2:23 -> 4:0'])
})

it('components have source maps', async () => {
  let config = {
    content: [{ raw: `container` }],
  }

  let input = css`
    @tailwind components;
  `

  let result = await run(input, config)
  let { sources, annotations } = parseSourceMaps(result)

  // All CSS generated by Tailwind CSS should be annotated with source maps
  // And always be able to point to the original source file
  expect(sources).not.toContain('<no source>')
  expect(sources.length).toBe(1)

  expect(annotations).toMatchSnapshot()
})

it('source maps for layer rules are not rewritten to point to @tailwind directives', async () => {
  let config = {
    content: [{ raw: `font-normal foo hover:foo` }],
  }

  let utilitiesFile = postcss.parse(
    css`
      @tailwind utilities;
    `,
    { from: 'components.css', map: { prev: map } }
  )

  let mainCssFile = postcss.parse(
    css`
      @layer utilities {
        .foo {
          background-color: red;
        }
      }
    `,
    { from: 'input.css', map: { prev: map } }
  )

  // Just pretend that there's an @import in `mainCssFile` that imports the nodes from `utilitiesFile`
  let input = postcss.root({
    nodes: [...utilitiesFile.nodes, ...mainCssFile.nodes],
    source: mainCssFile.source,
  })

  let result = await run(input, config)

  let { sources, annotations } = parseSourceMaps(result)

  // All CSS generated by Tailwind CSS should be annotated with source maps
  // And always be able to point to the original source file
  expect(sources).not.toContain('<no source>')

  // And we should see that the source map for the layer rule is not rewritten
  // to point to the @tailwind directive but instead points to the original
  expect(sources.length).toBe(2)
  expect(sources).toEqual(['components.css', 'input.css'])

  expect(annotations).toMatchSnapshot()
})
