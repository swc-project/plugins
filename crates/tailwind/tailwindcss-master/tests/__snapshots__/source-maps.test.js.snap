// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apply generates source maps 1`] = `
Array [
  "2:4 -> 2:4",
  "3:6-27 -> 3:6-27",
  "4:6-33 -> 4:6-18",
  "4:6-33 -> 5:6-17",
  "4:6-33 -> 6:6-24",
  "4:6-33 -> 7:6-61",
  "5:4 -> 8:4",
  "7:4 -> 10:4",
  "8:6-39 -> 11:6-39",
  "9:6-31 -> 12:6-18",
  "9:6-31 -> 13:6-17",
  "9:6-31 -> 14:6-24",
  "9:6-31 -> 15:6-61",
  "10:4 -> 16:4",
  "13:6 -> 18:4",
  "13:6-29 -> 19:6-18",
  "13:6-29 -> 20:6-17",
  "13:6-29 -> 21:6-24",
  "13:6 -> 22:6",
  "13:29 -> 23:0",
]
`;

exports[`components have source maps 1`] = `
Array [
  "2:4 -> 1:0",
  "2:4 -> 2:4",
  "2:24 -> 3:0",
  "2:4 -> 4:0",
  "2:4 -> 5:4",
  "2:4 -> 6:8",
  "2:24 -> 7:4",
  "2:24 -> 8:0",
  "2:4 -> 9:0",
  "2:4 -> 10:4",
  "2:4 -> 11:8",
  "2:24 -> 12:4",
  "2:24 -> 13:0",
  "2:4 -> 14:0",
  "2:4 -> 15:4",
  "2:4 -> 16:8",
  "2:24 -> 17:4",
  "2:24 -> 18:0",
  "2:4 -> 19:0",
  "2:4 -> 20:4",
  "2:4 -> 21:8",
  "2:24 -> 22:4",
  "2:24 -> 23:0",
  "2:4 -> 24:0",
  "2:4 -> 25:4",
  "2:4 -> 26:8",
  "2:24 -> 27:4",
  "2:24 -> 28:0",
]
`;

exports[`preflight + base have source maps 1`] = `
Array [
  "2:4 -> 1:0",
  "2:18-4 -> 3:1-2",
  "2:18 -> 6:1",
  "2:4 -> 8:0",
  "2:4-18 -> 11:2-32",
  "2:4-18 -> 12:2-25",
  "2:4-18 -> 13:2-29",
  "2:4-18 -> 14:2-31",
  "2:18 -> 15:0",
  "2:4 -> 17:0",
  "2:4-18 -> 19:2-18",
  "2:18 -> 20:0",
  "2:4 -> 22:0",
  "2:18 -> 28:1",
  "2:4 -> 30:0",
  "2:4-18 -> 31:2-26",
  "2:4-18 -> 32:2-40",
  "2:4-18 -> 33:2-26",
  "2:4-18 -> 34:2-21",
  "2:4-18 -> 35:2-230",
  "2:4-18 -> 36:2-39",
  "2:18 -> 37:0",
  "2:4 -> 39:0",
  "2:18 -> 42:1",
  "2:4 -> 44:0",
  "2:4-18 -> 45:2-19",
  "2:4-18 -> 46:2-30",
  "2:18 -> 47:0",
  "2:4 -> 49:0",
  "2:18 -> 53:1",
  "2:4 -> 55:0",
  "2:4-18 -> 56:2-19",
  "2:4-18 -> 57:2-24",
  "2:4-18 -> 58:2-31",
  "2:18 -> 59:0",
  "2:4 -> 61:0",
  "2:18 -> 63:1",
  "2:4 -> 65:0",
  "2:4-18 -> 66:2-35",
  "2:18 -> 67:0",
  "2:4 -> 69:0",
  "2:18 -> 71:1",
  "2:4 -> 73:0",
  "2:4-18 -> 79:2-20",
  "2:4-18 -> 80:2-22",
  "2:18 -> 81:0",
  "2:4 -> 83:0",
  "2:18 -> 85:1",
  "2:4 -> 87:0",
  "2:4-18 -> 88:2-16",
  "2:4-18 -> 89:2-26",
  "2:18 -> 90:0",
  "2:4 -> 92:0",
  "2:18 -> 94:1",
  "2:4 -> 96:0",
  "2:4-18 -> 98:2-21",
  "2:18 -> 99:0",
  "2:4 -> 101:0",
  "2:18 -> 104:1",
  "2:4 -> 106:0",
  "2:4-18 -> 110:2-121",
  "2:4-18 -> 111:2-24",
  "2:18 -> 112:0",
  "2:4 -> 114:0",
  "2:18 -> 116:1",
  "2:4 -> 118:0",
  "2:4-18 -> 119:2-16",
  "2:18 -> 120:0",
  "2:4 -> 122:0",
  "2:18 -> 124:1",
  "2:4 -> 126:0",
  "2:4-18 -> 128:2-16",
  "2:4-18 -> 129:2-16",
  "2:4-18 -> 130:2-20",
  "2:4-18 -> 131:2-26",
  "2:18 -> 132:0",
  "2:4 -> 134:0",
  "2:4-18 -> 135:2-17",
  "2:18 -> 136:0",
  "2:4 -> 138:0",
  "2:4-18 -> 139:2-13",
  "2:18 -> 140:0",
  "2:4 -> 142:0",
  "2:18 -> 146:1",
  "2:4 -> 148:0",
  "2:4-18 -> 149:2-24",
  "2:4-18 -> 150:2-31",
  "2:4-18 -> 151:2-35",
  "2:18 -> 152:0",
  "2:4 -> 154:0",
  "2:18 -> 158:1",
  "2:4 -> 160:0",
  "2:4-18 -> 165:2-30",
  "2:4-18 -> 166:2-25",
  "2:4-18 -> 167:2-30",
  "2:4-18 -> 168:2-30",
  "2:4-18 -> 169:2-24",
  "2:4-18 -> 170:2-19",
  "2:4-18 -> 171:2-20",
  "2:18 -> 172:0",
  "2:4 -> 174:0",
  "2:18 -> 176:1",
  "2:4 -> 178:0",
  "2:4-18 -> 180:2-22",
  "2:18 -> 181:0",
  "2:4 -> 183:0",
  "2:18 -> 186:1",
  "2:4 -> 188:0",
  "2:4-18 -> 192:2-36",
  "2:4-18 -> 193:2-39",
  "2:4-18 -> 194:2-32",
  "2:18 -> 195:0",
  "2:4 -> 197:0",
  "2:18 -> 199:1",
  "2:4 -> 201:0",
  "2:4-18 -> 202:2-15",
  "2:18 -> 203:0",
  "2:4 -> 205:0",
  "2:18 -> 207:1",
  "2:4 -> 209:0",
  "2:4-18 -> 210:2-18",
  "2:18 -> 211:0",
  "2:4 -> 213:0",
  "2:18 -> 215:1",
  "2:4 -> 217:0",
  "2:4-18 -> 218:2-26",
  "2:18 -> 219:0",
  "2:4 -> 221:0",
  "2:18 -> 223:1",
  "2:4 -> 225:0",
  "2:4-18 -> 227:2-14",
  "2:18 -> 228:0",
  "2:4 -> 230:0",
  "2:18 -> 233:1",
  "2:4 -> 235:0",
  "2:4-18 -> 236:2-39",
  "2:4-18 -> 237:2-30",
  "2:18 -> 238:0",
  "2:4 -> 240:0",
  "2:18 -> 242:1",
  "2:4 -> 244:0",
  "2:4-18 -> 245:2-26",
  "2:18 -> 246:0",
  "2:4 -> 248:0",
  "2:18 -> 251:1",
  "2:4 -> 253:0",
  "2:4-18 -> 254:2-36",
  "2:4-18 -> 255:2-23",
  "2:18 -> 256:0",
  "2:4 -> 258:0",
  "2:18 -> 260:1",
  "2:4 -> 262:0",
  "2:4-18 -> 263:2-20",
  "2:18 -> 264:0",
  "2:4 -> 266:0",
  "2:18 -> 268:1",
  "2:4 -> 270:0",
  "2:4-18 -> 283:2-11",
  "2:18 -> 284:0",
  "2:4 -> 286:0",
  "2:4-18 -> 287:2-11",
  "2:4-18 -> 288:2-12",
  "2:18 -> 289:0",
  "2:4 -> 291:0",
  "2:4-18 -> 292:2-12",
  "2:18 -> 293:0",
  "2:4 -> 295:0",
  "2:4-18 -> 298:2-18",
  "2:4-18 -> 299:2-11",
  "2:4-18 -> 300:2-12",
  "2:18 -> 301:0",
  "2:4 -> 303:0",
  "2:18 -> 305:1",
  "2:4 -> 307:0",
  "2:4-18 -> 308:2-18",
  "2:18 -> 309:0",
  "2:4 -> 311:0",
  "2:18 -> 314:1",
  "2:4 -> 316:0",
  "2:4-18 -> 318:2-20",
  "2:4-18 -> 319:2-24",
  "2:18 -> 320:0",
  "2:4 -> 322:0",
  "2:18 -> 324:1",
  "2:4 -> 326:0",
  "2:4-18 -> 328:2-17",
  "2:18 -> 329:0",
  "2:4 -> 331:0",
  "2:18 -> 333:1",
  "2:4 -> 334:0",
  "2:4-18 -> 335:2-17",
  "2:18 -> 336:0",
  "2:4 -> 338:0",
  "2:18 -> 342:1",
  "2:4 -> 344:0",
  "2:4-18 -> 352:2-24",
  "2:4-18 -> 353:2-32",
  "2:18 -> 354:0",
  "2:4 -> 356:0",
  "2:18 -> 358:1",
  "2:4 -> 360:0",
  "2:4-18 -> 362:2-17",
  "2:4-18 -> 363:2-14",
  "2:18 -> 364:0",
  "2:4-18 -> 366:0-72",
  "2:4 -> 367:0",
  "2:4-18 -> 368:2-15",
  "2:18 -> 369:0",
  "2:4 -> 371:0",
  "2:4-18 -> 372:2-26",
  "2:4-18 -> 373:2-26",
  "2:4-18 -> 374:2-21",
  "2:4-18 -> 375:2-21",
  "2:4-18 -> 376:2-16",
  "2:4-18 -> 377:2-16",
  "2:4-18 -> 378:2-16",
  "2:4-18 -> 379:2-17",
  "2:4-18 -> 380:2-17",
  "2:4-18 -> 381:2-15",
  "2:4-18 -> 382:2-15",
  "2:4-18 -> 383:2-20",
  "2:4-18 -> 384:2-40",
  "2:4-18 -> 385:2-17",
  "2:4-18 -> 386:2-22",
  "2:4-18 -> 387:2-24",
  "2:4-18 -> 388:2-25",
  "2:4-18 -> 389:2-26",
  "2:4-18 -> 390:2-20",
  "2:4-18 -> 391:2-29",
  "2:4-18 -> 392:2-30",
  "2:4-18 -> 393:2-40",
  "2:4-18 -> 394:2-36",
  "2:4-18 -> 395:2-29",
  "2:4-18 -> 396:2-24",
  "2:4-18 -> 397:2-32",
  "2:4-18 -> 398:2-14",
  "2:4-18 -> 399:2-20",
  "2:4-18 -> 400:2-18",
  "2:4-18 -> 401:2-19",
  "2:4-18 -> 402:2-20",
  "2:4-18 -> 403:2-16",
  "2:4-18 -> 404:2-18",
  "2:4-18 -> 405:2-15",
  "2:4-18 -> 406:2-21",
  "2:4-18 -> 407:2-23",
  "2:4-18 -> 408:2-29",
  "2:4-18 -> 409:2-27",
  "2:4-18 -> 410:2-28",
  "2:4-18 -> 411:2-29",
  "2:4-18 -> 412:2-25",
  "2:4-18 -> 413:2-26",
  "2:4-18 -> 414:2-27",
  "2:4 -> 415:2",
  "2:18 -> 416:0",
  "2:4 -> 418:0",
  "2:4-18 -> 419:2-26",
  "2:4-18 -> 420:2-26",
  "2:4-18 -> 421:2-21",
  "2:4-18 -> 422:2-21",
  "2:4-18 -> 423:2-16",
  "2:4-18 -> 424:2-16",
  "2:4-18 -> 425:2-16",
  "2:4-18 -> 426:2-17",
  "2:4-18 -> 427:2-17",
  "2:4-18 -> 428:2-15",
  "2:4-18 -> 429:2-15",
  "2:4-18 -> 430:2-20",
  "2:4-18 -> 431:2-40",
  "2:4-18 -> 432:2-17",
  "2:4-18 -> 433:2-22",
  "2:4-18 -> 434:2-24",
  "2:4-18 -> 435:2-25",
  "2:4-18 -> 436:2-26",
  "2:4-18 -> 437:2-20",
  "2:4-18 -> 438:2-29",
  "2:4-18 -> 439:2-30",
  "2:4-18 -> 440:2-40",
  "2:4-18 -> 441:2-36",
  "2:4-18 -> 442:2-29",
  "2:4-18 -> 443:2-24",
  "2:4-18 -> 444:2-32",
  "2:4-18 -> 445:2-14",
  "2:4-18 -> 446:2-20",
  "2:4-18 -> 447:2-18",
  "2:4-18 -> 448:2-19",
  "2:4-18 -> 449:2-20",
  "2:4-18 -> 450:2-16",
  "2:4-18 -> 451:2-18",
  "2:4-18 -> 452:2-15",
  "2:4-18 -> 453:2-21",
  "2:4-18 -> 454:2-23",
  "2:4-18 -> 455:2-29",
  "2:4-18 -> 456:2-27",
  "2:4-18 -> 457:2-28",
  "2:4-18 -> 458:2-29",
  "2:4-18 -> 459:2-25",
  "2:4-18 -> 460:2-26",
  "2:4-18 -> 461:2-27",
  "2:4 -> 462:2",
  "2:18 -> 463:0",
]
`;

exports[`source maps for layer rules are not rewritten to point to @tailwind directives 1`] = `
Array [
  "2:6 -> 1:0",
  "2:6 -> 2:10",
  "2:25 -> 3:0",
  "3:8 -> 4:8",
  "4:10-31 -> 5:10-31",
  "5:8 -> 6:8",
  "3:8 -> 7:8",
  "4:10-31 -> 8:10-31",
  "5:8 -> 9:8",
]
`;
